prov/gni/src/gnix_mbox_allocator.c

```diff
@@ -0,0 +1,537 @@
+		if (strstr(line, "hugetlbfs")) {
+			field = strtok(line, " ");
+			field = strtok(NULL, " ");
+
+			/*
+			 * Check if page is the correct size.
+			 */
+			if (statfs(field, &pg_size) == 0) {
+				if (pg_size.f_bsize == page_size) {
+					bytes_read = FI_SUCCESS;
+					break;
+				}
+			}
+		}
+	}
+
+	if (bytes_read == FI_SUCCESS)
+		*directory = strdup(field);
+
+	free(line);
+err1:
+	fclose(fd);
+err:
+	return bytes_read;
+}
+
+static int __generate_file_name(size_t page_size, char **filename)
+{
+	static const char basename[] = "gnix_map";
+	char *full_filename = NULL;
+	static int file_id;
+	char *huge_page = NULL;
+	int size;
+	int ret;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	if (!filename) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "filename pointer is NULL.\n");
+		ret = -FI_EINVAL;
+		goto err;
+	}
+
+	ret = __find_huge_page(page_size, &huge_page);
+	if (ret < 0) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Find huge page returned negative condition.\n");
+		goto err;
+	}
+
+	size = snprintf(NULL, 0, "%s/%s%d", huge_page, basename, file_id);
+	if (size < 0) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Error while gathering size for snprintf: %s\n",
+			  strerror(errno));
+		goto err1;
+	}
+
+	full_filename = malloc(size + 1);
+	if (!full_filename) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Error allocating full_filename: %s\n",
+			  strerror(errno));
+		ret = -FI_ENOMEM;
+		goto err1;
+	}
+
+	sprintf(full_filename, "%s/%s%d", huge_page, basename, file_id);
+
+	GNIX_DEBUG(FI_LOG_EP_CTRL, "Generated filename: %s\n", full_filename);
+
+	*filename = full_filename;
+
+	file_id++;
+
+err1:
+	free(huge_page);
+err:
+	return ret;
+}
+
+static int __open_huge_page(struct gnix_mbox_alloc_handle *handle)
+{
+	char *filename = NULL;
+	int ret;
+	int fd;
+
+	ret = __generate_file_name(handle->page_size, &filename);
+	if (ret < 0) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "Error in generating file name.");
+		goto err;
+	}
+
+	fd = open(filename, O_CREAT | O_RDWR | O_EXCL, 0700);
+	if (fd < 0) {
+		GNIX_WARN(FI_LOG_EP_CTRL, "IO Error: %s\n", strerror(errno));
+		ret = -FI_EIO;
+		goto err1;
+	}
+
+	handle->fd = fd;
+	handle->filename = filename;
+
+	GNIX_DEBUG(FI_LOG_EP_CTRL,
+		   "Successfully opened: %s with handle : %d\n.",
+		   handle->filename, handle->fd);
+
+	unlink(handle->filename);
+
+	return ret;
+
+err1:
+	free(filename);
+err:
+	return ret;
+}
+
+static size_t __page_count(struct gnix_mbox_alloc_handle *handle)
+{
+	size_t total_size = CEILING((handle->mbox_size * handle->mpmmap),
+				     handle->page_size);
+	size_t page_count;
+
+	page_count = total_size / handle->page_size;
+
+	GNIX_DEBUG(FI_LOG_EP_CTRL,
+		   "Mbox_size: %zu, mpmmap: %zu, page_size: %zu\n",
+		   handle->mbox_size, handle->mpmmap, handle->page_size);
+
+	GNIX_DEBUG(FI_LOG_EP_CTRL,
+		   "Total size: %zu, page_count: %zu\n", total_size,
+		   page_count);
+
+	if (page_count <= 0) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid size requested, truncating to single page.\n");
+		page_count = 1.0;
```

Comment:
Oops, that function used to deal with doubles, but I changed it. Forgot to fix that. Will update in round of updates.